maxima/g0/array_average/array_average.mac

```diff
@@ -0,0 +1,129 @@
+load("modal-basis")$
+load("out-scripts")$
+load("utilities_gyrokinetic")$
+
+writeCIncrExprsNoExpand1atomicadd(lhs, rhs) := block([expr],
+  expr : float(rhs),
+  for i : 1 thru length(expr) do (
+    if expr[i] # 0.0 then printf(fh, "  atomicAdd(&~a, ~a); ~%", lhs[i-1], expr[i])
+  )
+)$
+
+array_average(polyOrder_max,basisType,cfname,hfname) := block(
+
+  [headerf,fh,dims,single_elements,double_elements,triple_elements,
+  all_elements,intdims,dimNames,vars,basis,outVars,outNames,
+  intdim_s,dm_s,funcName,int_dim,b_reduced,f_e,w_e,f_c],
+
+  headerf : openw(hfname),
+  /* write prologue of header file*/
+  printf(headerf,"#pragma once~%~%"),
+  printf(headerf,"#include <gkyl_util.h>~%"),
+  printf(headerf,"EXTERN_C_BEG~%~%"),
+
+  fh : openw(cfname),
+  /* write prologue of c file*/
+  printf(fh,"#include <gkyl_array_average_kernels.h>~%~%"),
+
+  for polyOrder : 1 thru polyOrder_max do(
+    for dim : 1 thru 3 do (
+      dims : makelist(k, k, 1, dim),
+
+      /* Generate all single elements */
+      single_elements : dims,
+
+      /* Generate all pairs (combinations of size 2) */
+      double_elements : if length(dims) > 1 then 
+      create_list([dims[i], dims[j]], i, 1, length(dims), j, i+1, length(dims)) 
+      else [],
+
+      /* Generate all triplets (combinations of size 3) */
+      triple_elements : if length(dims) > 2 then 
+      create_list([dims[i], dims[j], dims[k]], i, 1, length(dims), j, i+1, length(dims), k, j+1, length(dims)) 
+      else [],
+
+      /* Combine single, double, and triple elements */
+      all_elements : append(
+        append(create_list([x], x, single_elements), double_elements),
+        triple_elements
+      ),
+
+      for i : 1 thru length(all_elements) do (
+        intdims : all_elements[i],
+
+        if dim = 3 then (dimNames : ["x","y","z"]),
+        if dim = 2 then (dimNames : ["x","y"]),
+        if dim = 1 then (dimNames : ["x"]),
+
+        /* Original basis */
+        [vars, basis] : loadBasis(basisType, dim, polyOrder),
+
+        outVars  : vars,
+        outNames : dimNames,
+
+        for j : 1 thru length(intdims) do(
+          outNames : delete(dimNames[intdims[j]],outNames),
+          outVars  : delete(vars[intdims[j]],outVars)
+        ),
+
+        intdim_s : "",
+        for j : 1 thru length(all_elements[i]) do(
+          intdim_s : sconcat(intdim_s,dimNames[intdims[j]])
+        ),
+
```

Comment:
Your comments elsewhere for what this script is doing is good, but can you clarify what these string concatenations are for? Just because some of these string concatenations look like short hand for things I usually painstakingly write out and so if you've figured out a more succinct was to do this, it would be good for a little bit of extra comment clarity.

Reply:
Hi,

I just want to comment on caching length(all_elements[i]). In the code above are recalculations, and it is bad practice, especially for a large length.
https://maxima.sourceforge.io/docs/manual/maxima_22.html

Reply:
> Hi,
> 
> I just want to comment on caching length(all_elements[i]). In the code above are recalculations, and it is bad practice, especially for a large length. https://maxima.sourceforge.io/docs/manual/maxima_22.html

Thank you for the advice. By the way, who are you? (:

Reply:
@Antoinehoff My name is Rubén Vega, I am a mathematician (bachelor).